src/v/cluster/types.cc

```diff
@@ -293,24 +318,188 @@ std::ostream& operator<<(
+struct topic_configuration_wrapper {
+    using value_t = topic_configuration_wrapper;
+    static constexpr serde::version_t redpanda_serde_version = 0;
+    static constexpr serde::version_t redpanda_serde_compat_version = 0;
+    std::optional<cluster::topic_configuration> cfg;
+
+    void serde_read(iobuf_parser& in, const serde::header& hdr) {
+        auto ns = serde::read_nested<ss::sstring>(in, hdr._bytes_left_limit);
+        auto tp = serde::read_nested<ss::sstring>(in, hdr._bytes_left_limit);
+
+        auto partition_count = serde::read_nested<int32_t>(
+          in, hdr._bytes_left_limit);
+        auto replication_factor = serde::read_nested<int16_t>(
+          in, hdr._bytes_left_limit);
+
+        cfg = cluster::topic_configuration(
+          model::ns(ns), model::topic(tp), partition_count, replication_factor);
+
+        cfg->properties.compression
+          = serde::read_nested<std::optional<model::compression>>(
+            in, hdr._bytes_left_limit);
+        cfg->properties.cleanup_policy_bitflags
+          = serde::read_nested<std::optional<model::cleanup_policy_bitflags>>(
+            in, hdr._bytes_left_limit);
+        cfg->properties.compaction_strategy
+          = serde::read_nested<std::optional<model::compaction_strategy>>(
+            in, hdr._bytes_left_limit);
+        cfg->properties.timestamp_type
+          = serde::read_nested<std::optional<model::timestamp_type>>(
+            in, hdr._bytes_left_limit);
+        cfg->properties.segment_size
+          = serde::read_nested<std::optional<size_t>>(
+            in, hdr._bytes_left_limit);
+        cfg->properties.retention_bytes = serde::read_nested<tristate<size_t>>(
+          in, hdr._bytes_left_limit);
+        cfg->properties.retention_duration
+          = serde::read_nested<tristate<std::chrono::milliseconds>>(
+            in, hdr._bytes_left_limit);
+        cfg->properties.recovery = serde::read_nested<std::optional<bool>>(
+          in, hdr._bytes_left_limit);
+        cfg->properties.shadow_indexing
+          = serde::read_nested<std::optional<cluster::shadow_indexing_mode>>(
+            in, hdr._bytes_left_limit);
+    }
+
+    void serde_write(iobuf& out) {
+        serde::write(out, cfg->tp_ns.ns());
+        serde::write(out, cfg->tp_ns.tp());
+        serde::write(out, cfg->partition_count);
+        serde::write(out, cfg->replication_factor);
+        serde::write(out, cfg->properties.compression);
+        serde::write(out, cfg->properties.cleanup_policy_bitflags);
+        serde::write(out, cfg->properties.compaction_strategy);
+        serde::write(out, cfg->properties.timestamp_type);
+        serde::write(out, cfg->properties.segment_size);
+        serde::write(out, cfg->properties.retention_bytes);
+        serde::write(out, cfg->properties.retention_duration);
+        serde::write(out, cfg->properties.recovery);
+        serde::write(out, cfg->properties.shadow_indexing);
```

Comment:
Can we split this commit? Since this is the first use case of serde, it would be nice to have a reference PR to start off with that shows the progressive change to serde for some example.

I could see three small commits:

1. Add the peek ahead trick with docs on how that works, and assert(false) for newer versions
2. Add support for existing fields in the new serde version
3. Add the two new fields

Reply:
since the way serde is dealing with enums is going to change pretty soon I decided to switch to adl for now
it will be trivial to switch back to serde for the `topic_configuation` struct because we already have a version

I addressed all other comments